### PR TITLE
[clang][HLSL] Add radians intrinsic

### LINKED ISSUE #99151: Implement the `radians` HLSL Function
- [ ] Implement `radians` clang builtin,
- [ ] Link `radians` clang builtin with `hlsl_intrinsics.h`
- [ ] Add sema checks for `radians` to `CheckHLSLBuiltinFunctionCall` in `SemaChecking.cpp`
- [ ] Add codegen for `radians` to `EmitHLSLBuiltinExpr` in `CGBuiltin.cpp`
- [ ] Add codegen tests to `clang/test/CodeGenHLSL/builtins/radians.hlsl`
- [ ] Add sema tests to `clang/test/SemaHLSL/BuiltIns/radians-errors.hlsl`
- [ ] Create the `int_spv_radians` intrinsic in `IntrinsicsSPIRV.td`
- [ ] In SPIRVInstructionSelector.cpp create the `radians` lowering and map  it to `int_spv_radians` in `SPIRVInstructionSelector::selectIntrinsic`.
- [ ] Create SPIR-V backend test case in `llvm/test/CodeGen/SPIRV/hlsl-intrinsics/radians.ll`

## DirectX

There were no DXIL opcodes found for `radians`.

## SPIR-V

# Radians:

## Description:
**Radians**  
  
Converts *degrees* to radians, i.e., *degrees* \* π / 180.  
  
The operand *degrees* must be a scalar or vector whose component type is
16-bit or 32-bit floating-point.  
  
*Result Type* and the type of *degrees* must be the same type. Results
are computed per component.

<table>
<colgroup>
<col style="width: 20%" />
<col style="width: 20%" />
<col style="width: 20%" />
<col style="width: 20%" />
<col style="width: 20%" />
</colgroup>
<thead>
<tr>
<th>Number</th>
<th>Operand 1</th>
<th>Operand 2</th>
<th>Operand 3</th>
<th>Operand 4</th>
</tr>
</thead>
<tbody>
<tr>
<td class="tableblock halign-left valign-top"><p>11</p></td>
<td
class="tableblock halign-left valign-top"><p><em>&lt;id&gt;</em><br />
<em>degrees</em></p></td>
<td></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>

### Example 1
```hlsl
//dxc radians_test.hlsl -T lib_6_8 -enable-16bit-types -O0

export float4 fn(float4 p1) {
    return radians(p1);
}
```
## HLSL:

Converts the specified value from degrees to radians.

| *ret* radians(*x*) |
|--------------------|

 

## Parameters

| Item                                                   | Description                            |
|--------------------------------------------------------|----------------------------------------|
| <span id="x"></span><span id="X"></span>*x*<br/> | \[in\] The specified value.<br/> |

 

## Return Value

The *x* parameter converted from degrees to radians.

## Type Description

| Name  | [**Template Type**](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-intrinsic-functions.md)                                                  | [**Component Type**](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-intrinsic-functions.md) | Size                           |
|-------|----------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------|--------------------------------|
| *x*   | [**scalar**](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-intrinsic-functions.md), **vector**, or **matrix** | [**float**](/windows/desktop/WinProg/windows-data-types)                        | any                            |
| *ret* | same as input *x*                                                                                              | [**float**](/windows/desktop/WinProg/windows-data-types)                        | same dimension(s) as input *x* |

 

## Minimum Shader Model

This function is supported in the following shader models.

| Shader Model                                                                       | Supported |
|------------------------------------------------------------------------------------|-----------|
| [Shader Model 1 (DirectX HLSL)](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-sm1.md) and higher shader models | yes       |

 

## See also

<dl> <dt>

[**Intrinsic Functions (DirectX HLSL)**](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-intrinsic-functions.md)
</dt> </dl>

### PR DESCRIPTION
**Pull request by adam-yang**
partially fixes #99151

### Changes
* Implemented `radians` clang builtin
* Linked `radians` clang builtin with `hlsl_intrinsics.h`
* Added sema checks for `radians` to `CheckHLSLBuiltinFunctionCall` in `SemaChecking.cpp`
* Add codegen for `radians` to `EmitHLSLBuiltinExpr` in `CGBuiltin.cpp`
* Add codegen tests to `clang/test/CodeGenHLSL/builtins/radians.hlsl`
* Add sema tests to `clang/test/SemaHLSL/BuiltIns/radians-errors.hlsl`

### Related PRs
* [[DXIL] Add radians intrinsic #110616](https://github.com/llvm/llvm-project/pull/110616)
* [[SPIRV] Add radians intrinsic #110800](https://github.com/llvm/llvm-project/pull/110800)

### CODE REVIEW

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 1]
```diff
@@ -18852,6 +18852,14 @@ case Builtin::BI__builtin_hlsl_elementwise_isinf: {
         retType, CGM.getHLSLRuntime().getSignIntrinsic(),
         ArrayRef<Value *>{Op0}, nullptr, "hlsl.sign");
   }
+  case Builtin::BI__builtin_hlsl_elementwise_radians: {
+    Value *Op0 = EmitScalarExpr(E->getArg(0));
+    if (!E->getArg(0)->getType()->hasFloatingRepresentation())
```
Comment (farzonl) [inline clang/lib/CodeGen/CGBuiltin.cpp:18857]: Line 1864 of  `SemaHLSL.cpp` you have a check for this, Turn this into an assert.

File: clang/test/SemaHLSL/BuiltIns/radians-errors.hlsl [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+float builtin_radians_int_to_float_promotion(int p1) {
+  return __builtin_hlsl_elementwise_radians(p1);
+  // expected-error@-1 {{passing 'int' to parameter of incompatible type 'float'}}
+}
+
+float2 builtin_radians_int2_to_float2_promotion(int2 p1) {
+  return __builtin_hlsl_elementwise_radians(p1);
+  // expected-error@-1 {{passing 'int2' (aka 'vector<int, 2>') to parameter of incompatible type '__attribute__((__vector_size__(2 * sizeof(float)))) float' (vector of 2 'float' values)}}
+}
+
```
Comment (farzonl) [inline clang/test/SemaHLSL/BuiltIns/radians-errors.hlsl:27]: There is a file called `clang\test\SemaHLSL\BuiltIns\half-float-only-errors.hlsl` Add `__builtin_hlsl_elementwise_radians` to that file.

**Review by farzonl — APPROVED** [review]

**Review by farzonl — CHANGES_REQUESTED** [review]
Please run clang-format

**Review by spall — APPROVED** [review]
LGTM but I didn't look too closely at the tests.

**Comment by adam-yang** [comment]
@farzonl This is also ready for merging if there's no further feedback.

**Review by farzonl — APPROVED** [review]

**Pull request merged**